src/System.Runtime.Analyzers/Core/SpecifyIFormatProvider.cs

```diff
@@ -27,42 +34,203 @@ public abstract class SpecifyIFormatProviderAnalyzer : DiagnosticAnalyzer
+                var currentCultureProperty = CultureInfoType?.GetMembers("CurrentCulture").OfType<IPropertySymbol>().SingleOrDefault();
+                var invariantCultureProperty = CultureInfoType?.GetMembers("InvariantCulture").OfType<IPropertySymbol>().SingleOrDefault();
+                var currentUICultureProperty = CultureInfoType?.GetMembers("CurrentUICulture").OfType<IPropertySymbol>().SingleOrDefault();
+                var installedUICultureProperty = CultureInfoType?.GetMembers("InstalledUICulture").OfType<IPropertySymbol>().SingleOrDefault();
+
+                var threadType = csaContext.Compilation.GetTypeByMetadataName("System.Threading.Thread");
+                var currentThreadCurrentUICultureProperty = threadType?.GetMembers("CurrentUICulture").OfType<IPropertySymbol>().SingleOrDefault();
+
+                var activatorType = csaContext.Compilation.GetTypeByMetadataName("System.Activator");
+                var resourceManagerType = csaContext.Compilation.GetTypeByMetadataName("System.Resources.ResourceManager");
+
+                var computerInfoType = csaContext.Compilation.GetTypeByMetadataName("Microsoft.VisualBasic.Devices.ComputerInfo");
+                var installedUICulturePropertyOfComputerInfoType = computerInfoType?.GetMembers("InstalledUICulture").OfType<IPropertySymbol>().SingleOrDefault();
+                #endregion
+
+                csaContext.RegisterOperationAction(oaContext =>
+                {
+                    var invocationExpression = (IInvocationExpression)oaContext.Operation;
+                    var targetMethod = invocationExpression.TargetMethod;
+
+                    #region "Exceptions"
+                    if (targetMethod.IsGenericMethod ||
+                        (targetMethod.ContainingType != null && 
+                         (activatorType != null && activatorType.Equals(targetMethod.ContainingType)) ||
+                         (resourceManagerType != null && resourceManagerType.Equals(targetMethod.ContainingType))))
+                    {
+                        return;
+                    }
+                    #endregion
+
+                    #region "IFormatProviderAlternateStringRule Only"
+                    if (stringType != null && CultureInfoType != null &&
+                        (targetMethod.Equals(stringFormatMemberWithStringAndObjectParameter) ||
+                         targetMethod.Equals(stringFormatMemberWithStringObjectAndObjectParameter) ||
+                         targetMethod.Equals(stringFormatMemberWithStringObjectObjectAndObjectParameter) ||
+                         targetMethod.Equals(stringFormatMemberWithStringAndParamsObjectParameter)))
+                    {
+                        // Sample message for IFormatProviderAlternateStringRule: Because the behavior of string.Format(string, object) could vary based on the current user's locale settings,
```

Comment:
These messages are wordy but they are actionable providing information on what a replacement could be based on the user scenario.